In [163]:
from __future__ import unicode_literals

# Importing the libraries
import numpy as np
import pandas as pd

# from sklearn import preprocessing
from sklearn.preprocessing import Imputer

# 1. Reading our pandas dataframes inputs variables

We read our 28 weeks life plants input datasets **Phreatic Level, Precipitations, Temperature, Luminosity, Direction and Wind Speed** 

## 1.1. Wind Direction - Training dataset 

In [164]:
wind_direction_training = pd.read_csv('../data/processed/WindDirection/28-weeks_January-12_August-03_2018/' \
                                 'Wind-Direction-TRAINING_January-12_August-03_2018_DummiesValues.csv', )

In [165]:
print(wind_direction_training.shape)
wind_direction_training.head(10)

(7801, 8)


E  N  NE  NO  O  S  SE  SO
0  1  0   0   0  0  0   0   0
1  0  0   0   1  0  0   0   0
2  0  0   0   0  0  0   0   1
3  1  0   0   0  0  0   0   0
4  0  0   0   0  0  0   1   0
5  0  1   0   0  0  0   0   0
6  0  0   0   1  0  0   0   0
7  0  0   0   1  0  0   0   0
8  0  0   0   0  0  0   1   0
9  0  0   0   1  0  0   0   0

## 1.2. Luminosity - Training dataset 

In [166]:
luminosity_training = pd.read_csv('../data/processed/Luminosity/28-weeks_January-12_August-03_2018/' \
                                 'Luminosity_Normalized_TRAINING_January-12_August-03.csv', )

In [167]:
print(luminosity_training.shape)
luminosity_training.head(10)

(7801, 1)


Luz (lux)
0   0.476135
1   0.000000
2   0.000000
3   0.816777
4   0.645003
5   0.492464
6   0.000090
7   0.000000
8   0.000000
9   0.000000

## 1.3. Phreatic Level - Training dataset 

In [168]:
phreatic_level_training = pd.read_csv('../data/processed/PhreaticLevel/28-weeks_January-12_August-03_2018/' \
                                 'Phreatic-Level_Normalized_TRAINING_January-12_August-03.csv', )

In [169]:
print(phreatic_level_training.shape)
phreatic_level_training.head(10)

(3904, 25)


PORVL2N1  PORVL2N2  PORVL4N1  PORVL5N1  PORVL6N1  PORVL7N1  PORVL8N1  \
0  0.481707  0.810976  0.805031  0.780488  0.743056  1.000000  0.936709   
1  0.390244  0.810976  0.685535  0.609756  0.618056  0.816456  0.816456   
2  0.304878  0.518293  0.534591  0.329268  0.451389  0.626582  0.645570   
3  0.030488  0.000000  0.264151  0.128049  0.173611  0.265823  0.348101   
4  0.298780  0.359756  0.427673  0.262195  0.840278  0.658228  0.658228   
5  0.402439  0.481707  0.496855  0.426829  0.937500  0.556962  0.601266   
6  0.579268  0.493902  0.616352  0.530488  0.993056  0.784810  0.740506   
7  0.439024  0.810976  0.754717  0.695122  0.687500  0.898734  0.879747   
8  0.487805  0.810976  0.798742  0.774390  0.743056  1.000000  0.943038   
9  0.512195  0.573171  0.572327  0.500000  1.000000  0.721519  0.708861   

   PORVL9N1  PORVL10N1  PORVL13N1    ...      PORVL18N2  PORVL18N3  PORVL18N4  \
0  1.000000    1.00000   0.827160    ...       0.914110   0.971631   1.000000   
1  0.853659    1.00000   0.746914    ...       0.815951   0.829787   0.685185   
2  0.664634    0.56250   0.604938    ...       0.484663   0.709220   0.453704   
3  0.439024    0.37500   0.296296    ...       0.190184   0.460993   0.231481   
4  0.481707    0.53125   0.518519    ...       0.478528   0.780142   0.564815   
5  0.518293    0.51875   0.543210    ...       0.496933   0.687943   0.490741   
6  0.560976    0.71250   0.703704    ...       0.754601   0.822695   0.787037   
7  0.926829    1.00000   0.820988    ...       0.901840   0.886525   0.777778   
8  1.000000    1.00000   0.820988    ...       0.907975   0.971631   1.000000   
9  0.560976    0.72500   0.666667    ...       0.625767   0.723404   0.731481   

   PORVL21N1  PORVL21N2  PORVL21N3  PORVL21N4  PORVL21N5  PORVL24N1  PORVL24N2  
0   1.000000   1.000000   1.000000   1.000000   0.825806   0.897436   1.000000  
1   0.992537   0.949686   1.000000   0.891892   0.825806   0.685897   1.000000  
2   0.895522   0.685535   0.591463   0.547297   0.554839   0.538462   0.721805  
3   0.589552   0.477987   0.378049   0.229730   0.335484   0.038462   0.308271  
4   0.492537   0.421384   0.420732   0.547297   0.567742   0.391026   0.661654  
5   0.500000   0.522013   0.567073   0.506757   0.677419   0.429487   0.601504  
6   0.597015   0.490566   0.615854   0.493243   0.754839   0.666667   0.819549  
7   1.000000   1.000000   1.000000   1.000000   0.819355   0.769231   1.000000  
8   1.000000   1.000000   1.000000   1.000000   0.825806   0.897436   1.000000  
9   0.604478   0.433962   0.621951   0.229730   0.709677   0.698718   0.759398  

[10 rows x 25 columns]

## 1.4. Precipitations - Training dataset 

In [170]:
precipitations_training = pd.read_csv('../data/processed/Precipitations/28-weeks_January-12_August-03_2018/' \
                                 'Precipitations_Normalized_TRAINING_January-12_August-03.csv', )

In [171]:
print(precipitations_training.shape)
precipitations_training.head(15)

(7800, 1)


Precipitación (ml)
0             0.000000
1             0.000000
2             0.000000
3             0.000000
4             0.000000
5             0.000000
6             0.000000
7             0.000000
8             0.000000
9             0.000000
10            0.000000
11            0.000000
12            0.000000
13            0.013182
14            0.000000

## 1.5. Temperatura - Training dataset 

In [172]:
temperature_training = pd.read_csv('../data/processed/Temperature/28-weeks_January-12_August-03_2018/' \
                                 'Temperature_Normalized_TRAINING_January-12_August-03.csv', )

In [173]:
print(temperature_training.shape)
temperature_training.head(10)

(7801, 1)


Temperatura (C)
0         0.486758
1         0.185381
2         0.519597
3         0.155720
4         0.592691
5         0.552966
6         0.129237
7         0.238347
8         0.701801
9         0.179025

## 1.6. Wind Speed - Training dataset 

In [174]:
wind_speed_training = pd.read_csv('../data/processed/WindSpeed/28-weeks_January-12_August-03_2018/' \
                                 'Wind-Speed_Normalized_TRAINING_January-12_August-03.csv', )

In [175]:
print(wind_speed_training.shape)
wind_speed_training.head(10)

(7801, 1)


Velocidad del viento (Km/h)
0                     0.000000
1                     0.000000
2                     0.030568
3                     0.071186
4                     0.142857
5                     0.000000
6                     0.122807
7                     0.000000
8                     0.022105
9                     0.000000

## 1.7. Banana - Cluster weight 

Cluster weight training data of one week of productivity taken as a baseline is loaded.

The week is from July 27th to August 3rd

In [176]:
racimitos_weight_training = pd.read_csv('../data/processed/Cultivos/UltimosRacimos/' \
                                 'racimitos-weight_Normalized_TRAINING_July-27_August-03.csv', )

In [177]:
print(racimitos_weight_training.shape)
racimitos_weight_training.head(10)

(6116, 1)


pesoRacimo
0    0.493524
1    0.207810
2    0.131619
3    0.398286
4    0.388762
5    0.407810
6    0.379238
7    0.426857
8    0.179238
9    0.255429

# 2. Forming the whole Sioma training dataset - January 12th to August 3rd - All input variables

We will make a single dataset where all the standardized training data are from:

- Wind Direction
- Luminosity
- Precipitations
- Temperature
- Phreatic Level
- Wind Speed
- Cluster weight of the week from July 27th to August 3rd.

In [178]:
sioma_28_weeks = pd.concat([luminosity_training, precipitations_training, temperature_training, wind_speed_training, wind_direction_training, phreatic_level_training, racimitos_weight_training], axis=1)
sioma_28_weeks

Luz (lux)  Precipitación (ml)  Temperatura (C)  \
0      0.476135            0.000000         0.486758   
1      0.000000            0.000000         0.185381   
2      0.000000            0.000000         0.519597   
3      0.816777            0.000000         0.155720   
4      0.645003            0.000000         0.592691   
5      0.492464            0.000000         0.552966   
6      0.000090            0.000000         0.129237   
7      0.000000            0.000000         0.238347   
8      0.000000            0.000000         0.701801   
9      0.000000            0.000000         0.179025   
10     0.200323            0.000000         0.552966   
11     0.538812            0.000000         0.486758   
12     0.000000            0.000000         0.155720   
13     0.359465            0.013182         0.251589   
14     0.476135            0.000000         0.496822   
15     0.738166            0.000000         0.235169   
16     0.106729            0.026364         0.397246   
17     0.789682            0.000000         0.112818   
18     0.000090            0.000000         0.486758   
19     0.705706            0.000000         0.271716   
20     0.000000            0.000000         0.029661   
21     0.000000            0.000000         0.261653   
22     0.005778            0.000000         0.221928   
23     0.411376            0.000000         0.284958   
24     0.371793            0.000000         0.711864   
25     0.000000            0.000000         0.086335   
26     0.430307            0.000000         0.731462   
27     0.000000            0.000000         0.354343   
28     0.000000            0.000000         0.179025   
29     0.934757            0.000000         0.245233   
...         ...                 ...              ...   
7771   0.015073            0.000000         0.645657   
7772   0.000000            0.000000         0.288136   
7773   0.179024            0.000000         0.185381   
7774   0.000072            0.000000         0.095869   
7775   0.000000            0.000000         0.241525   
7776   0.000000            0.000000         0.453390   
7777   0.000000            0.000000         0.238347   
7778   0.000000            0.000000         0.112818   
7779   0.351462            0.000000         0.503178   
7780   0.000000            0.000000         0.450212   
7781   0.000000            0.000000         0.457097   
7782   0.000000            0.000000         0.115996   
7783   0.460327            0.000000         0.384004   
7784   0.355446            0.000000         0.367585   
7785   0.000000            0.000000         0.682203   
7786   0.000000            0.000000         0.344280   
7787   0.000000            0.000000         0.182203   
7788   0.000108            0.000000         0.208686   
7789   0.000000            0.006591         0.390890   
7790   0.221658            0.000000         0.638771   
7791   0.602925            0.000000         0.493114   
7792   0.864005            0.000000         0.142479   
7793   0.129194            0.000000         0.179025   
7794   0.000000            0.000000         0.095869   
7795   0.000000            0.000000         0.609110   
7796   0.000000            0.000000         0.142479   
7797   0.371793            0.000000         0.357521   
7798   0.000000            0.000000         0.698623   
7799   0.000072            0.000000         0.208686   
7800   0.913978                 NaN         0.619174   

      Velocidad del viento (Km/h)  E  N  NE  NO  O  S     ...      PORVL18N3  \
0                        0.000000  1  0   0   0  0  0     ...       0.971631   
1                        0.000000  0  0   0   1  0  0     ...       0.829787   
2                        0.030568  0  0   0   0  0  0     ...       0.709220   
3                        0.071186  1  0   0   0  0  0     ...       0.460993   
4                        0.142857  0  0   0   0  0  0     ...       0.780142   
5                        0.000000  0  1   0   0  0  0

## 2.1 Filling **`NaN:`** values
---

We evaluate if `sioma_28_weeks` dataframe has null type `NaN`

Have null values.

We see that many columns have remained with null values, product of the concatenation of these different dataframes, since not all had the same
number of records `X[i]`

We proceed to fill in that missing data
A more common solution alternative is to take the average of the column values where the data is missing, in this case find the average of all the columns that our data set has `sioma_28_weeks`

As we said before, we will use as a strategy, fill in the missing data with the average of the other values in the column where data is missing.

axis is the axis along which we are going to do the Imputer operation, by default it is equal to 0, which means that this operation will be along the columns

In [179]:
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)

# We adjust the values transforming the columns where you need data 
# with the fit() method. 
# We pass the dataframe sioma_28_weeks and 
# with the variables where I can find the empty values, in this case 
# all variables [::] and fit 
imputer = imputer.fit(sioma_28_weeks[::])
sioma_training_full_values = imputer.transform(sioma_28_weeks[::])

/home/bgarcial/anaconda3/lib/python3.7/site-packages/sklearn/utils/deprecation.py:58: DeprecationWarning: Class Imputer is deprecated; Imputer was deprecated in version 0.20 and will be removed in 0.22. Import impute.SimpleImputer from sklearn instead.
  warnings.warn(msg, category=DeprecationWarning)


The previous step generate the ** `sioma_training_full_values` ** array which don't have null values

In [180]:
sioma_training_full_values

array([[4.76134936e-01, 0.00000000e+00, 4.86758475e-01, ...,
        8.97435897e-01, 1.00000000e+00, 4.93523810e-01],
       [0.00000000e+00, 0.00000000e+00, 1.85381356e-01, ...,
        6.85897436e-01, 1.00000000e+00, 2.07809524e-01],
       [0.00000000e+00, 0.00000000e+00, 5.19597458e-01, ...,
        5.38461538e-01, 7.21804511e-01, 1.31619048e-01],
       ...,
       [0.00000000e+00, 0.00000000e+00, 6.98622881e-01, ...,
        6.14936473e-01, 7.77499027e-01, 3.57563425e-01],
       [7.17746277e-05, 0.00000000e+00, 2.08686441e-01, ...,
        6.14936473e-01, 7.77499027e-01, 3.57563425e-01],
       [9.13978109e-01, 2.84053030e-03, 6.19173729e-01, ...,
        6.14936473e-01, 7.77499027e-01, 3.57563425e-01]])

In [181]:
print(sioma_training_full_values.shape)
np.isnan(sioma_training_full_values)

(7801, 38)


array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

In [182]:
cols = list(sioma_28_weeks.columns)
print(sioma_28_weeks.shape)
print(sioma_training_full_values.shape)
print(cols)

(7801, 38)
(7801, 38)
['Luz (lux)', 'Precipitación (ml)', 'Temperatura (C)', 'Velocidad del viento (Km/h)', 'E', 'N', 'NE', 'NO', 'O', 'S', 'SE', 'SO', 'PORVL2N1', 'PORVL2N2', 'PORVL4N1', 'PORVL5N1', 'PORVL6N1', 'PORVL7N1', 'PORVL8N1', 'PORVL9N1', 'PORVL10N1', 'PORVL13N1', 'PORVL14N1', 'PORVL15N1', 'PORVL16N1', 'PORVL16N2', 'PORVL18N1', 'PORVL18N2', 'PORVL18N3', 'PORVL18N4', 'PORVL21N1', 'PORVL21N2', 'PORVL21N3', 'PORVL21N4', 'PORVL21N5', 'PORVL24N1', 'PORVL24N2', 'pesoRacimo']


In [183]:
# Then, the `sioma_training_full_values` don't have missing values
np.isnan(sioma_training_full_values)

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

In [184]:
# We turn sioma_training_full_values to dataframe
sioma_28_weeks_training_df = pd.DataFrame(data=sioma_training_full_values, columns=cols)
print(sioma_28_weeks_training_df.isnull().any())


Luz (lux)                      False
Precipitación (ml)             False
Temperatura (C)                False
Velocidad del viento (Km/h)    False
E                              False
N                              False
NE                             False
NO                             False
O                              False
S                              False
SE                             False
SO                             False
PORVL2N1                       False
PORVL2N2                       False
PORVL4N1                       False
PORVL5N1                       False
PORVL6N1                       False
PORVL7N1                       False
PORVL8N1                       False
PORVL9N1                       False
PORVL10N1                      False
PORVL13N1                      False
PORVL14N1                      False
PORVL15N1                      False
PORVL16N1                      False
PORVL16N2                      False
PORVL18N1                      False
P

In [185]:
# Mostramos todos los datos y vemos que desde el registro 6616 todos los datos son iguales porque muchos de ellos
# antes eran valores nulos que fueron llenados con el promedio de los otros y dieron igual valor, escogemos solo los diferentes.

In [186]:
sioma_28_weeks_training_df

Luz (lux)  Precipitación (ml)  Temperatura (C)  \
0      0.476135            0.000000         0.486758   
1      0.000000            0.000000         0.185381   
2      0.000000            0.000000         0.519597   
3      0.816777            0.000000         0.155720   
4      0.645003            0.000000         0.592691   
5      0.492464            0.000000         0.552966   
6      0.000090            0.000000         0.129237   
7      0.000000            0.000000         0.238347   
8      0.000000            0.000000         0.701801   
9      0.000000            0.000000         0.179025   
10     0.200323            0.000000         0.552966   
11     0.538812            0.000000         0.486758   
12     0.000000            0.000000         0.155720   
13     0.359465            0.013182         0.251589   
14     0.476135            0.000000         0.496822   
15     0.738166            0.000000         0.235169   
16     0.106729            0.026364         0.397246   
17     0.789682            0.000000         0.112818   
18     0.000090            0.000000         0.486758   
19     0.705706            0.000000         0.271716   
20     0.000000            0.000000         0.029661   
21     0.000000            0.000000         0.261653   
22     0.005778            0.000000         0.221928   
23     0.411376            0.000000         0.284958   
24     0.371793            0.000000         0.711864   
25     0.000000            0.000000         0.086335   
26     0.430307            0.000000         0.731462   
27     0.000000            0.000000         0.354343   
28     0.000000            0.000000         0.179025   
29     0.934757            0.000000         0.245233   
...         ...                 ...              ...   
7771   0.015073            0.000000         0.645657   
7772   0.000000            0.000000         0.288136   
7773   0.179024            0.000000         0.185381   
7774   0.000072            0.000000         0.095869   
7775   0.000000            0.000000         0.241525   
7776   0.000000            0.000000         0.453390   
7777   0.000000            0.000000         0.238347   
7778   0.000000            0.000000         0.112818   
7779   0.351462            0.000000         0.503178   
7780   0.000000            0.000000         0.450212   
7781   0.000000            0.000000         0.457097   
7782   0.000000            0.000000         0.115996   
7783   0.460327            0.000000         0.384004   
7784   0.355446            0.000000         0.367585   
7785   0.000000            0.000000         0.682203   
7786   0.000000            0.000000         0.344280   
7787   0.000000            0.000000         0.182203   
7788   0.000108            0.000000         0.208686   
7789   0.000000            0.006591         0.390890   
7790   0.221658            0.000000         0.638771   
7791   0.602925            0.000000         0.493114   
7792   0.864005            0.000000         0.142479   
7793   0.129194            0.000000         0.179025   
7794   0.000000            0.000000         0.095869   
7795   0.000000            0.000000         0.609110   
7796   0.000000            0.000000         0.142479   
7797   0.371793            0.000000         0.357521   
7798   0.000000            0.000000         0.698623   
7799   0.000072            0.000000         0.208686   
7800   0.913978            0.002841         0.619174   

      Velocidad del viento (Km/h)    E    N   NE   NO    O    S     ...      \
0                        0.000000  1.0  0.0  0.0  0.0  0.0  0.0     ...       
1                        0.000000  0.0  0.0  0.0  1.0  0.0  0.0     ...       
2                        0.030568  0.0  0.0  0.0  0.0  0.0  0.0     ...       
3                        0.071186  1.0  0.0  0.0  0.0  0.0  0.0     ...       
4                        0.142857  0.0  0.0  0.0  0.0  0.0  0.0     ...       
5                        0.000000  0.0  1.0  0.0  0.0  0.0 

In [187]:
sioma_28_weeks_training_df.shape

(7801, 38)

In [188]:
# data.drop_duplicates(subset=['pesoRacimo'], keep="first")
#sioma_28_weeks_training_df.drop_duplicates(subset=['pesoRacimo'], keep='first', inplace=False)


In [189]:
sioma_28_weeks_training_df.to_csv('../data/processed/sioma-inputs-output-unified_training_28_weeks.csv', sep=',', header=True, index=False)
#sioma_28_weeks_training_df.shape

We have in a single dataset a total of 7801 training data of:

- Wind Direction
- Luminosity
- Precipitations
- Temperature
- Phreatic Level
- Wind Speed
- Cluster weight of the week from July 27th to August 3rd.